app/javascript/packs/simple-form-validation.js

```diff
@@ -0,0 +1,67 @@
+import { loadPolyfills } from '@18f/identity-polyfill';
+
+/** @typedef {{t:(key:string)=>string, key:(key:string)=>string}} LoginGovI18n */
+/** @typedef {{LoginGov:{I18n:LoginGovI18n}}} LoginGovGlobal */
+
+const { I18n } = /** @type {typeof window & LoginGovGlobal}} */ (window).LoginGov;
+
+const PATTERN_TYPES = ['dob', 'personal-key', 'ssn', 'state_id_number', 'zipcode'];
+
+/**
+ * Given a submit event, disables all submit buttons within the target form.
+ *
+ * @param {Event} event Submit event.
+ */
+function disableFormSubmit(event) {
+  const form = /** @type {HTMLFormElement} */ (event.target);
+  [...form.querySelectorAll('[type="submit"]')].forEach((submit) => {
+    /** @type {HTMLInputElement|HTMLButtonElement} */ (submit).disabled = true;
+  });
+}
+
+/**
+ * Given an `input` or `invalid` event, updates custom validity of the given input.
+ *
+ * @param {Event} event Input or invalid event.
+ */
+function checkInputValidity(event) {
+  const input = /** @type {HTMLInputElement} */ (event.target);
+  input.setCustomValidity('');
+
+  if (input.validity.valueMissing) {
+    input.setCustomValidity(I18n.t('simple_form.required.text'));
+  } else if (input.validity.patternMismatch) {
+    PATTERN_TYPES.forEach((type) => {
+      if (input.classList.contains(type)) {
+        input.setCustomValidity(I18n.t(`idv.errors.pattern_mismatch.${I18n.key(type)}`));
+      }
+    });
+  }
+}
+
+/**
+ * Binds validation to a given input.
+ *
+ * @param {HTMLInputElement} input Input element.
+ */
+function validateInput(input) {
+  input.addEventListener('input', checkInputValidity);
+  input.addEventListener('invalid', checkInputValidity);
+}
+
+/**
+ * Initializes validation on a form element.
+ *
+ * @param {HTMLFormElement} form Form to initialize.
+ */
+export function initialize(form) {
+  /** @type {HTMLInputElement[]} */ ([...form.querySelectorAll('.field')]).forEach(validateInput);
+  form.addEventListener('submit', disableFormSubmit);
+}
+
+loadPolyfills(['classlist']).then(() => {
+  if (document.currentScript) {
+    const form = /** @type {HTMLFormElement} */ (document.currentScript.previousElementSibling);
+    initialize(form);
```

Comment:
If I'm reading this right, this relies on the `idp_form_for` specific ordering. WDYT of adding like a `data-` attribute via our helper method? A slightly more explicit link between the two files? This would break if we chose to wrap the forms in `<div>` or something, right?

Reply:
> If I'm reading this right, this relies on the `idp_form_for` specific ordering. WDYT of adding like a `data-` attribute via our helper method? A slightly more explicit link between the two files? This would break if we chose to wrap the forms in `<div>` or something, right?

Yeah, you're reading it right. This relies on (and takes advantage of!) the placement of the script being immediately following the form element.

I had seen a few advantages with this:

- If there are multiple forms on a page, it "just works", where each form adds one copy of the script, and that copy of the script will only enhance the form for which the script had been added.
- If we're okay to make this assumption of the order of `form` + `script`, we don't have to mess about with `DOMContentLoaded` or other async/deferred script execution.

To the first point, I had also been wondering about some sort of "add script once" or `content_for` to deduplicate scripts where we might expect multiple copies. I could see this working a bit better with your idea of an explicit attribute, though it might be a bit more work and complexity to set this up.

To your question about breakage if we change wrappers:

- Would it really be expected that `form_for` would return anything other than a `<form>` element?
- Interestingly enough, it wouldn't actually break, due to the fact that we're merely relying on `querySelector` and `submit` events, where technically if there were a `div` wrapper, it would still receive the bubbling `submit` event ([example](https://codepen.io/aduth/pen/qBZJaag)).

So I don't know that I'm particularly troubled by it, though it could be justified if we go the route of trying to memoize these scripts. I also plan to add some tests for this script, which might help protect against such hypothetical future breakage?

Reply:
Re: wrappers, unsure if we'd add another div for layouts, styling etc (all hypothetical), that was my thinking

But yeah if there was a test or something that like helps enforce the expectation that the elements are adjacent that would make me feel better about relying on that dependency